src/model/schema/ThunderstoreSchema.ts

```diff
@@ -0,0 +1,131 @@
+import R2Error from "../errors/R2Error";
+
+// Disabled for now, will become relevant when we migrate away from JSON modules.
+export interface Schema {
+    schema_version: string,
+    games: [string:SchemaEntry],
+    communities: [string:ThunderstoreEntry],
+    packageInstallers: [],
+    modloaderPackages: [ModloaderPackage],
+}
+
+export interface SchemaEntry {
+    uuid: string,
+    label: string,
+    meta: EntryMeta,
+    distributions: [EntryDist],
+    thunderstore?: ThunderstoreEntry,
+    r2modman?: EntryR2,
+}
+
+export interface ThunderstoreEntry {
+    displayName: string,
+    discordUrl?: string,
+    wikiUrl?: string,
+    categories: any,
+    sections: any,
+}
+
+export interface ModloaderPackage {
+    identifier: string,
+    rootFolder: string,
+    variant: string,
+}
+
+export interface EntryMeta {
+    displayName: string,
+    iconUrl: string,
+}
+
+export interface EntryDist {
+    platform: DistributionPlatform,
+    identifier?: string,
+}
+
+export interface EntryR2 {
+    internalFolderName: string,
+    dataFolderName?: string,
+    settingsIdentifier: string,
+    packageIndex: string,
+    exclusionsUrl: string,
+    steamFolderName: string,
+    exeNames: [string],
+    gameInstanceType: GameInstanceType,
+    gameSelectionDisplayMode: GameSelectionDisplayMode,
+    packageLoader: string,
+    additionalSearchStrings: [string],
+    installRules: [InstallRule],
+}
+
+export interface InstallRule {
+    route: string,
+    defaultFileExtensions: [string],
+    trackingMethod: string,
+    subRoutes: [InstallRule],
+    isDefaultLocation: boolean,
+}
+
+enum GameInstanceType {
+    GAME = "game",
+    SERVER = "server",
+}
+
+enum DistributionPlatform {
+    STEAM = "steam",
+    STEAM_DIRECT = "steam-direct",
+    EGS = "epic-games-store",
+    OCULUS = "oculus",
+    ORIGIN = "origin",
+    XBOX_GAME_PASS = "xbox-game-pass",
+    OTHER = "other",
+}
+
+export class EcosystemSchema {
+    private static _isValidated: boolean = false;
+
+    /**
+     * Get a validated instance of the ecosystem schema.
+     */
+    static get get() {
+        if (this._isValidated) {
+            return schema;
+        }
+
+        // Validate the schema via its schema schema.
+        const ajv = new Ajv();
+        addFormats(ajv);
+
+        const validate = ajv.compile(schemaSchema);
+        const isOk = validate(schema);
+
+        if (!isOk) {
+            throw new R2Error("Schema validation error", ajv.errorsText(validate.errors));
+        }
+
+        this._isValidated = true;
+        return schema;
+    }
+
+    /**
+     * Get a list of games which have valid r2mm entries.
+     */
+    static get r2mm_games() {
+        const schema = this.get!;
+
+        const games = [];
+        for (const jsonObj in schema.games) {
+            const name: keyof typeof schema.games = jsonObj as any;
+            games.push(schema.games[name]);
+        }
+        return games.filter((x) => x.r2modman != null);
+    }
```

Comment:
I'm not exactly sure what's the intention here? Does this somehow differ from simple

```typescript
static get r2mm_games() {
    return Object.values(this.get.games).filter((x) => x.r2modman != null);
}

At least `const jsonObj in schema.games` is confusing as `jsonObj` is the string key, not serialized JSON.
```

Reply:
Thanks, that works much better. I had a sticking point early on and I decided to hit it with a sledgehammer.

Reply:
IDK what changed between the review and current version, but now I **do** get compilation errors when ever I try to access the `r2modman` property without `!`. I think something like this might fix it here in the getter:

```typescript
static get supportedGames() {
    return Object.values(this.get.games).filter(
        (x): x is typeof x & {r2modman: NonNullable<typeof x["r2modman"]>} => x.r2modman != null
    );
}
```